File: lib/features/analytics/domain/aggregators.dart
import '../../../core/date_utils.dart';
import 'analytics_input.dart';
import 'analytics_range.dart';
import 'series.dart';

/// How a bucket combines the days inside it.
enum BucketReducer {
  /// Mean of the days that have data. Right for rates -- calories per day,
  /// hours slept, goal score. A week bucket showing 15,000 kcal is unreadable.
  mean,

  /// Sum over the bucket. Right for quantities -- training volume, session
  /// counts, where the weekly total is the number that means something.
  sum,
}

/// Folds a per-day map into one point per bucket.
///
/// A bucket with no observed day yields a null point, not a zero, so the chart
/// can draw a gap and the averages can ignore it.
MetricSeries bucketize(
  Map<DateTime, double> byDay,
  DateRange range,
  AnalyticsBucket bucket,
  BucketReducer reducer,
) {
  final buckets = <DateTime, List<double>>{};
  for (final day in range.days) {
    final start = bucketStartFor(day, bucket);
    buckets.putIfAbsent(start, () => <double>[]);
    final value = byDay[day];
    if (value != null) buckets[start]!.add(value);
  }

  return MetricSeries([
    for (final start in bucketStarts(range, bucket))
      TimeSeriesPoint(start, _reduce(buckets[start] ?? const [], reducer)),
  ]);
}

double? _reduce(List<double> values, BucketReducer reducer) {
  if (values.isEmpty) return null;
  final total = values.reduce((a, b) => a + b);
  return reducer == BucketReducer.sum ? total : total / values.length;
}

/// Per-day nutrition, keyed by calendar day. Unlogged days are absent, so they
/// stay distinguishable from days that were logged as zero.
Map<DateTime, double> nutritionByDay(
  List<DailyNutrition> nutrition,
  double Function(DailyNutrition) pick,
) =>
    {
      for (final day in nutrition)
        if (day.logged) AppDateUtils.intToDate(day.dateInt): pick(day),
    };

/// Total training volume per day. Days without a session are absent.
Map<DateTime, double> volumeByDay(List<TrainingSession> sessions) {
  final out = <DateTime, double>{};
  for (final session in sessions) {
    out[session.day] = (out[session.day] ?? 0) + session.volume;
  }
  return out;
}

/// Completed sessions per day.
Map<DateTime, double> sessionCountByDay(List<TrainingSession> sessions) {
  final out = <DateTime, double>{};
  for (final session in sessions) {
    out[session.day] = (out[session.day] ?? 0) + 1;
  }
  return out;
}

/// Minutes trained per day.
Map<DateTime, double> trainingMinutesByDay(List<TrainingSession> sessions) {
  final out = <DateTime, double>{};
  for (final session in sessions) {
    out[session.day] = (out[session.day] ?? 0) + session.duration.inMinutes;
  }
  return out;
}

Map<DateTime, double> sleepHoursByDay(List<SleepNight> nights) => {
      for (final night in nights)
        AppDateUtils.startOfDay(night.day): night.hours
    };

/// Body weight per day.
///
/// Never bucketed by the caller: weigh-ins are sparse and irregular, and
/// averaging them into weekly bars throws away the exact shape the EMA exists
/// to smooth. The daily series carries gaps and the chart connects across them.
Map<DateTime, double> weightByDay(List<WeighIn> weighIns) =>
    {for (final w in weighIns) AppDateUtils.startOfDay(w.day): w.kg};

/// Sets performed per primary muscle, for the distribution bars.
///
/// Exercises with no `primaryMuscle` are grouped under [unlabelledMuscle]
/// rather than dropped -- a user's own untagged exercise still represents real
/// work, and silently omitting it would understate their total.
const String unlabelledMuscle = 'other';

Map<String, int> setsPerMuscle(
  List<TrainingSession> sessions,
  Map<String, ExerciseRef> exercises,
) {
  final out = <String, int>{};
  for (final session in sessions) {
    for (final set in session.sets) {
      final muscle = exercises[set.exerciseId]?.primaryMuscle?.trim();
      final key =
          (muscle == null || muscle.isEmpty) ? unlabelledMuscle : muscle;
      out[key] = (out[key] ?? 0) + 1;
    }
  }
  return out;
}

/// Standard deviation of bedtime, in hours -- the consistency figure.
///
/// Bedtimes are projected onto a continuous scale before averaging, otherwise
/// 23:30 and 00:30 average to noon instead of midnight and every user looks
/// wildly inconsistent.
double? bedtimeConsistencyHours(List<SleepNight> nights) {
  if (nights.length < 2) return null;
  final samples = nights.map((n) {
    final h = n.startedAt.hour + n.startedAt.minute / 60.0;
    return h < 12 ? h + 24 : h;
  }).toList();
  return standardDeviation(samples);
}

File: Packages/WellnessKit/Sources/WellnessDomain/Aggregators.swift
import Foundation
import WellnessModels

public enum BucketReducer: Sendable {
    case mean, sum
}

public enum Aggregators: Sendable {

    public static func bucketize(
        _ byDay: [Date: Double],
        range: DateRange,
        bucket: AnalyticsBucket,
        reducer: BucketReducer
    ) -> MetricSeries {
        var buckets: [Date: [Double]] = [:]
        for day in DateHelpers.daysIn(range) {
            let start = bucketStartFor(day, bucket)
            buckets[start, default: []].append(contentsOf:
                byDay[day].map { [$0] } ?? [])
        }
        let starts = bucketStarts(range, bucket)
        return MetricSeries(starts.map { start in
            TimeSeriesPoint(t: start, value: reduce(buckets[start] ?? [], reducer))
        })
    }

    private static func reduce(_ values: [Double], _ reducer: BucketReducer) -> Double? {
        guard !values.isEmpty else { return nil }
        let total = values.reduce(0, +)
        return reducer == .sum ? total : total / Double(values.count)
    }

    public static func nutritionByDay(
        _ nutrition: [DailyNutrition],
        pick: (DailyNutrition) -> Double
    ) -> [Date: Double] {
        var out: [Date: Double] = [:]
        for day in nutrition where day.logged {
            out[DateHelpers.intToDate(day.dateInt)] = pick(day)
        }
        return out
    }

    public static func volumeByDay(_ sessions: [TrainingSession]) -> [Date: Double] {
        var out: [Date: Double] = [:]
        for s in sessions {
            let d = DateHelpers.startOfDay(s.startedAt)
            out[d, default: 0] += s.volume
        }
        return out
    }

    public static func sessionCountByDay(_ sessions: [TrainingSession]) -> [Date: Double] {
        var out: [Date: Double] = [:]
        for s in sessions {
            out[DateHelpers.startOfDay(s.startedAt), default: 0] += 1
        }
        return out
    }

    public static func trainingMinutesByDay(_ sessions: [TrainingSession]) -> [Date: Double] {
        var out: [Date: Double] = [:]
        for s in sessions {
            out[DateHelpers.startOfDay(s.startedAt), default: 0] += s.duration / 60
        }
        return out
    }

    public static func sleepHoursByDay(_ nights: [SleepNight]) -> [Date: Double] {
        var out: [Date: Double] = [:]
        for n in nights { out[DateHelpers.startOfDay(n.day)] = n.hours }
        return out
    }

    public static func weightByDay(_ weighIns: [WeighIn]) -> [Date: Double] {
        var out: [Date: Double] = [:]
        for w in weighIns { out[DateHelpers.startOfDay(w.day)] = w.kg }
        return out
    }

    public static let unlabelledMuscle = "other"

    public static func setsPerMuscle(
        _ sessions: [TrainingSession],
        _ exercises: [String: ExerciseRef]
    ) -> [String: Int] {
        var out: [String: Int] = [:]
        for session in sessions {
            for s in session.sets {
                let muscle = exercises[s.exerciseId]?.primaryMuscle?.trimmingCharacters(in: .whitespaces)
                let key = (muscle == nil || muscle!.isEmpty) ? unlabelledMuscle : muscle!
                out[key, default: 0] += 1
            }
        }
        return out
    }

    public static func bedtimeConsistencyHours(_ nights: [SleepNight]) -> Double? {
        guard nights.count >= 2 else { return nil }
        let samples = nights.map { n -> Double in
            let cal = Calendar(identifier: .gregorian)
            let h = Double(cal.component(.hour, from: n.startedAt))
                + Double(cal.component(.minute, from: n.startedAt)) / 60
            return h < 12 ? h + 24 : h
        }
        return standardDeviation(samples)
    }
}

public func bucketStartFor(_ day: Date, _ bucket: AnalyticsBucket) -> Date {
    let d = DateHelpers.startOfDay(day)
    switch bucket {
    case .day:   return d
    case .week:  return DateHelpers.startOfWeek(d)
    case .month:
        let cal = Calendar(identifier: .gregorian)
        let c = cal.dateComponents([.year, .month], from: d)
        return cal.date(from: c)!
    }
}

public func bucketStarts(_ range: DateRange, _ bucket: AnalyticsBucket) -> [Date] {
    var seen = Set<Date>()
    var out: [Date] = []
    for day in DateHelpers.daysIn(range) {
        let start = bucketStartFor(day, bucket)
        if seen.insert(start).inserted { out.append(start) }
    }
    return out
}
